File: frontend/src/lib/config/channel.ts
export type ReleaseCapabilities = {
	releaseChannel: string;
	onlineRegistry: boolean;
	onlineTemplateStore: boolean;
	onlineSkillStore: boolean;
	templateAutoUpdate: boolean;
	firstLaunchBootstrap: boolean;
	appUpdateCheck: boolean;
	externalBrowserLinks: boolean;
	localTemplateImport: boolean;
	packagedRuntimes: boolean;
};

function envChannel(): string {
	const channel = import.meta.env.VITE_PRESTO_CHANNEL ?? 'slim';
	return channel === 'portable' ? 'portable' : 'slim';
}

function capabilitiesForChannel(channel: string): ReleaseCapabilities {
	if (channel === 'portable') {
		return {
			releaseChannel: 'portable',
			onlineRegistry: false,
			onlineTemplateStore: false,
			onlineSkillStore: false,
			templateAutoUpdate: false,
			firstLaunchBootstrap: false,
			appUpdateCheck: false,
			externalBrowserLinks: false,
			localTemplateImport: true,
			packagedRuntimes: true,
		};
	}
	return {
		releaseChannel: 'slim',
		onlineRegistry: true,
		onlineTemplateStore: true,
		onlineSkillStore: true,
		templateAutoUpdate: true,
		firstLaunchBootstrap: true,
		appUpdateCheck: true,
		externalBrowserLinks: true,
		localTemplateImport: true,
		packagedRuntimes: false,
	};
}

export const defaultCapabilities = capabilitiesForChannel(envChannel());

export async function loadCapabilities(): Promise<ReleaseCapabilities> {
	try {
		const capabilities = await window.go?.main?.App?.GetCapabilities?.();
		if (capabilities) {
			return capabilities;
		}
	} catch {
		// Browser/dev fallback is driven by VITE_PRESTO_CHANNEL.
	}
	return defaultCapabilities;
}

export function isPortable(capabilities: ReleaseCapabilities): boolean {
	return capabilities.releaseChannel === 'portable';
}

export function releaseChannelLabel(capabilities: ReleaseCapabilities): string {
	return isPortable(capabilities) ? '离线便携包' : '默认精剪包';
}

File: cmd/presto-desktop/channel.go
package main

import "github.com/mrered/presto/internal/api"

var releaseChannel = "slim"

type ReleaseCapabilities struct {
	ReleaseChannel       string `json:"releaseChannel"`
	OnlineRegistry       bool   `json:"onlineRegistry"`
	OnlineTemplateStore  bool   `json:"onlineTemplateStore"`
	OnlineSkillStore     bool   `json:"onlineSkillStore"`
	TemplateAutoUpdate   bool   `json:"templateAutoUpdate"`
	FirstLaunchBootstrap bool   `json:"firstLaunchBootstrap"`
	AppUpdateCheck       bool   `json:"appUpdateCheck"`
	ExternalBrowserLinks bool   `json:"externalBrowserLinks"`
	LocalTemplateImport  bool   `json:"localTemplateImport"`
	PackagedRuntimes     bool   `json:"packagedRuntimes"`
}

func currentReleaseCapabilities() ReleaseCapabilities {
	switch releaseChannel {
	case "portable":
		return ReleaseCapabilities{
			ReleaseChannel:       "portable",
			OnlineRegistry:       false,
			OnlineTemplateStore:  false,
			OnlineSkillStore:     false,
			TemplateAutoUpdate:   false,
			FirstLaunchBootstrap: false,
			AppUpdateCheck:       false,
			ExternalBrowserLinks: false,
			LocalTemplateImport:  true,
			PackagedRuntimes:     true,
		}
	default:
		return slimReleaseCapabilities()
	}
}

func slimReleaseCapabilities() ReleaseCapabilities {
	return ReleaseCapabilities{
		ReleaseChannel:       "slim",
		OnlineRegistry:       true,
		OnlineTemplateStore:  true,
		OnlineSkillStore:     true,
		TemplateAutoUpdate:   true,
		FirstLaunchBootstrap: true,
		AppUpdateCheck:       true,
		ExternalBrowserLinks: true,
		LocalTemplateImport:  true,
		PackagedRuntimes:     false,
	}
}

func normalizeReleaseCapabilities(capabilities ReleaseCapabilities) ReleaseCapabilities {
	if capabilities.ReleaseChannel == "" {
		return currentReleaseCapabilities()
	}
	return capabilities
}

func toAPIReleaseCapabilities(capabilities ReleaseCapabilities) api.ReleaseCapabilities {
	return api.ReleaseCapabilities{
		ReleaseChannel:       capabilities.ReleaseChannel,
		OnlineRegistry:       capabilities.OnlineRegistry,
		OnlineTemplateStore:  capabilities.OnlineTemplateStore,
		OnlineSkillStore:     capabilities.OnlineSkillStore,
		TemplateAutoUpdate:   capabilities.TemplateAutoUpdate,
		FirstLaunchBootstrap: capabilities.FirstLaunchBootstrap,
		AppUpdateCheck:       capabilities.AppUpdateCheck,
		ExternalBrowserLinks: capabilities.ExternalBrowserLinks,
		LocalTemplateImport:  capabilities.LocalTemplateImport,
		PackagedRuntimes:     capabilities.PackagedRuntimes,
	}
}
